src/main/kotlin/uk/gov/gdx/datashare/services/OutboundEventQueueService.kt

```diff
@@ -10,16 +21,240 @@ import uk.gov.gdx.datashare.queue.SqsProperties
+      .targetKeyId(keyId)
+      .build()
+    kmsClient.createAlias(aliasRequest)
+  }
+
+  private fun tagKey(keyId: String?) {
+    val tagRequest = TagResourceRequest.builder()
+      .keyId(keyId)
+      .tags(
+        Tag.builder().tagKey("Environment").tagValue(environment).build(),
+        Tag.builder().tagKey("Owner").tagValue("gdx-dev-team@digital.cabinet-office.gov.uk").build(),
+        Tag.builder().tagKey("Product").tagValue("Government Data Exchange").build(),
+        Tag.builder().tagKey("Repository").tagValue("https://github.com/alphagov/gdx-data-share-poc").build(),
+        Tag.builder().tagKey("Source").tagValue("Application").build(),
+      ).build()
+    kmsClient.tagResource(tagRequest)
+  }
+
+  private fun createQueue(
+    queueName: String,
+    keyId: String,
+    messageRetentionDays: Int,
+    dlqArn: String? = null,
+    acquirerPrincipal: String? = null,
+  ): String {
+    val createQueueRequest = CreateQueueRequest.builder()
+      .queueName(queueName)
+      .attributes(buildQueueAttributes(keyId, queueName, messageRetentionDays, dlqArn, acquirerPrincipal))
       .build()
-    queue.sqsClient.sendMessage(request)
+    val createQueueResponse = sqsClient.createQueue(createQueueRequest)
+    return createQueueResponse.queueUrl()
+  }
+
+  private fun buildQueueAttributes(
+    keyId: String,
+    queueName: String,
+    messageRetentionDays: Int,
+    dlqArn: String?,
+    acquirerPrincipal: String?,
+  ): HashMap<QueueAttributeName, String> {
+    val attributes = HashMap<QueueAttributeName, String>()
+    attributes[QueueAttributeName.KMS_MASTER_KEY_ID] = keyId
+    attributes[QueueAttributeName.KMS_DATA_KEY_REUSE_PERIOD_SECONDS] = "300" // 5 minutes
+    attributes[QueueAttributeName.MESSAGE_RETENTION_PERIOD] = "${messageRetentionDays * 24 * 60 * 60}"
+    attributes[QueueAttributeName.VISIBILITY_TIMEOUT] = "30"
+    attributes[QueueAttributeName.POLICY] = queuePolicy(acquirerPrincipal)
+    if (isFifoQueue(queueName)) {
+      attributes[QueueAttributeName.FIFO_QUEUE] = "true"
+    }
+    if (dlqArn != null) {
+      attributes[QueueAttributeName.REDRIVE_POLICY] = """{"maxReceiveCount": "10", "deadLetterTargetArn": "$dlqArn"}"""
+    }
+    return attributes
+  }
+
+  private fun getQueueArn(queueUrl: String): String? {
+    val request = GetQueueAttributesRequest.builder()
+      .queueUrl(queueUrl)
+      .attributeNames(QueueAttributeName.QUEUE_ARN)
+      .build()
+    val response = sqsClient.getQueueAttributes(request)
+    return response.attributes()[QueueAttributeName.QUEUE_ARN]
+  }
+
+  private fun isFifoQueue(queueName: String) = queueName.endsWith(".fifo")
+
+  private fun dlqName(queueName: String) =
+    if (isFifoQueue(queueName)) "${queueName.substringBefore(".fifo")}_dlq.fifo" else "${queueName}_dlq"
+
+  private fun keyPolicy(acquirerPrincipal: String?): String {
+    if (acquirerPrincipal == null) {
+      return """
+        {
```

Comment:
Is there a nicer way of doing this by stringifying a data class or something?

Reply:
It depends what you think is nicer. It think that would look something like 
```kotlin
@JsonInclude(JsonInclude.Include.NON_NULL)
data class QueuePolicy(
  private val Version: String = "2012-10-12",
  private val Id: String = "cross-account-access",
  private val Statement: List<QueuePolicyStatment>,
)

data class QueuePolicyStatment(
  private val Sid: String,
  private val Effect: String,
  private val Action: List<String>,
  private val Principal: PolicyPrincipal? = null,
  private val Condition: PolicyCondition? = null,
)

data class PolicyPrincipal(
  private val AWS: String,
)

data class PolicyCondition(
  private val Bool: BoolPolicyCondition,
)

data class BoolPolicyCondition(
  @JsonProperty("aws:SecureTransport")
  private val secureTransport: String
)

fun newQueuePolicy(acquirerPrincipal: String?): String {
  if (acquirerPrincipal == null) {
    return QueuePolicy(
      Statement = listOf(
        QueuePolicyStatment(
          Sid = "httpsonly",
          Effect = "Deny",
          Action = listOf("sqs:*"),
          Condition = PolicyCondition(
            BoolPolicyCondition(secureTransport = "false")
          )
        )
      )
    ).toJson()
  }
}
```
which is much less readable to my mind

Reply:
I thought there may be an existing type in the sdk, but if not I am happy to go with as is

Reply:
No types - everything just takes a string policy document.
```kotlin
mapOf(
  "Version" to "2012-10-12",
  "Id" to "cross-account-access",
  "Statement" to listOf(
    mapOf(
      "Sid" to "acquirerAccess",
      "Principal" to mapOf("AWS" to "$acquirerPrincpal"),
      "Effect" to "Allow",
      "Action" to listOf(
        "SQS:GetQueueUrl",
        "SQS:GetQueueAttributes",
        "SQS:ChangeMessageVisibility",
        "SQS:ReceiveMessage",
        "SQS:DeleteMessage",
      ),
    ),
    mapOf(
      "Sid" to "httponly",
      "Effect" to "Deny",
      "Action" to "sqs:*",
      "Condition" to mapOf("Bool" to mapOf("aws:SecureTransport" to "false"))
    ),
  )
)
```
is a little better than the above, but I think I still prefer the hand-rolled JSON